File: PYTHON/Pertemuan 01/ProseduralBermasalah.py
import sys
sys.stdout.reconfigure(encoding='utf-8')
# Variabel GLOBAL – bisa diakses fungsi mana pun
nama_mhs = []
ipk_mhs = []
jumlah_mhs = 0

def tambah_mahasiswa(nama: str, ipk: float):
    global jumlah_mhs
    nama_mhs.append(nama)
    ipk_mhs.append(ipk)   # nilai langsung disimpan, tidak ada validasi!
    jumlah_mhs += 1

def hack_ipk(index: int):
    ipk_mhs[index] = 4.0   # tidak ada yang mencegah!
    print(f"⚠  IPK indeks ke-{index} diubah paksa menjadi 4.0 oleh hack_ipk()!")

def tampilkan_semua_mahasiswa():
    print("\n── Daftar Mahasiswa (Pendekatan Prosedural) ──")
    for i in range(jumlah_mhs):
        print(f"  [{i}] Nama: {nama_mhs[i]:<15} | IPK: {ipk_mhs[i]:.2f}")

def main():
    print("╔══════════════════════════════════════════════╗")
    print("║   PENDEKATAN PROSEDURAL – Masalah Nyata      ║")
    print("╚══════════════════════════════════════════════╝\n")

    # Menambah data mahasiswa
    tambah_mahasiswa("Budi Santoso", 3.50)
    tambah_mahasiswa("Sari Wulandari", 3.75)
    tambah_mahasiswa("Andi Pratama", 2.90)

    print("Data awal setelah ditambahkan:")
    tampilkan_semua_mahasiswa()

    # ⚠ Masalah: fungsi lain bebas mengubah data langsung!
    print()
    hack_ipk(2)   # mengubah IPK Andi secara paksa

    print("\nData setelah hack_ipk() dipanggil:")
    tampilkan_semua_mahasiswa()

    # ⚠ Bahkan bisa diubah langsung dari luar fungsi!
    print("\n⚠  Mengubah IPK langsung via list global (indeks 0):")
    ipk_mhs[0] = -99.0   # nilai tidak masuk akal, tapi tidak ada yang mencegah
    tampilkan_semua_mahasiswa()

    print("\n──────────────────────────────────────────────────")
    print("KESIMPULAN: Tanpa enkapsulasi, data bisa diubah")
    print("           dari mana saja & kapan saja → BERBAHAYA!")
    print("──────────────────────────────────────────────────")

if __name__ == "__main__":
    main()

File: JAVA/Pertemuan 01/src/id/ac/uph/pbo/pertemuan01/ProseduralBermasalah.java
package id.ac.uph.pbo.pertemuan01;

public class ProseduralBermasalah {

    // ── Variabel GLOBAL – bisa diakses fungsi mana pun ──────────────────────
    static String[] namaMhs    = new String[100];
    static double[] ipkMhs     = new double[100];
    static int      jumlahMhs  = 0;

    // ── Fungsi untuk menambah mahasiswa (TANPA validasi IPK) ────────────────
    static void tambahMahasiswa(String nama, double ipk) {
        namaMhs[jumlahMhs] = nama;
        ipkMhs[jumlahMhs]  = ipk;   // nilai langsung disimpan, tidak ada validasi!
        jumlahMhs++;
    }

    // ── Fungsi yang bisa "membobol" data IPK secara sembarangan ─────────────
    static void hackIPK(int index) {
        ipkMhs[index] = 4.0;   // tidak ada yang mencegah!
        System.out.println("⚠  IPK indeks ke-" + index
                + " diubah paksa menjadi 4.0 oleh hackIPK()!");
    }

    // ── Tampilkan semua data ─────────────────────────────────────────────────
    static void tampilkanSemuaMahasiswa() {
        System.out.println("\n── Daftar Mahasiswa (Pendekatan Prosedural) ──");
        for (int i = 0; i < jumlahMhs; i++) {
            System.out.printf("  [%d] Nama: %-15s | IPK: %.2f%n",
                    i, namaMhs[i], ipkMhs[i]);
        }
    }

    // ── main: demonstrasi masalah ────────────────────────────────────────────
    public static void main(String[] args) {

        System.out.println("╔══════════════════════════════════════════════╗");
        System.out.println("║   PENDEKATAN PROSEDURAL – Masalah Nyata      ║");
        System.out.println("╚══════════════════════════════════════════════╝\n");

        // Menambah data mahasiswa
        tambahMahasiswa("Budi Santoso",  3.50);
        tambahMahasiswa("Sari Wulandari", 3.75);
        tambahMahasiswa("Andi Pratama",   2.90);

        System.out.println("Data awal setelah ditambahkan:");
        tampilkanSemuaMahasiswa();

        // ⚠ Masalah: fungsi lain bebas mengubah data langsung!
        System.out.println();
        hackIPK(2);   // mengubah IPK Andi secara paksa

        System.out.println("\nData setelah hackIPK() dipanggil:");
        tampilkanSemuaMahasiswa();

        // ⚠ Bahkan bisa diubah langsung dari luar fungsi!
        System.out.println("\n⚠  Mengubah IPK langsung via array global (indeks 0):");
        ipkMhs[0] = -99.0;   // nilai tidak masuk akal, tapi tidak ada yang mencegah
        tampilkanSemuaMahasiswa();

        System.out.println("\n──────────────────────────────────────────────────");
        System.out.println("KESIMPULAN: Tanpa enkapsulasi, data bisa diubah");
        System.out.println("           dari mana saja & kapan saja → BERBAHAYA!");
        System.out.println("──────────────────────────────────────────────────");
    }
}
